### PR TITLE
Install via conda

### LINKED ISSUE #225: How to install a stand-alone parcels package?
## Why?

At Geomar, we have a central Jupyter Hub allowing for running remote kernels on virtually all of our analysis machines. The remote backends rely on Conda environments. Ideally, we'd be able to include a parcels environment satisfying the `requirements.txt` and adding a parcels package.

The basic logic behind our backend setup follows a very nice blog post by @rabernat: https://medium.com/@rabernat/custom-conda-environments-for-data-science-on-hpc-clusters-32d58c63aa95

The following env file (largely based on parcel's `requirements.txt`) seems to get close to a working parcels setup:

```yml
name: py2_parcels
channels:
  - https://conda.anaconda.org/conda-forge
  - https://repo.continuum.io/pkgs/free
dependencies:
  - python=2.7
  - cachetools>=1.0.0
  - cgen
  - enum34
  - flake8>=2.1.0
  - jupyter
  - matplotlib
  - netcdf4>=1.1.9
  - numpy>=1.9.1
  - progressbar2
  - py>=1.4.27
  - pymbolic
  - python-dateutil
  - scipy>=0.16.0
  - six>=1.10.0
  - xarray>=0.5.1
  - pip:
      - pytest-ipynb
      - pytest>=2.7.0
      - "git+https://github.com/OceanParcels/parcels.git@v0.9"
```

## Problem

Running the [parcels tutorial](http://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/examples/parcels_tutorial.ipynb) with essentially the following steps works only partially:

```bash
source activate py2_parcels
git clone https://github.com/OceanParcels/parcels.git parcels
cd parcels
git fetch --tags
git checkout v0.9
python parcels/scripts/pull_data.py
jupyter notebook
```

I do, however, run into problems with `gcc` being unable to compile, because of a missing `includes/parcels.h`. Import statements seem to use the installed version (as intended) while only the tutorial itself and the `pull_data.py` from the cloned version are used.

## Going forward

I'd be fine with providing a PR for a decent `setup.py` and / or documentation. Before being able to do so, I think, I'll have to understand how a user is supposed to work with parcels:
- Keeping a full copy of the repo for each experiment?
- Or having a central installation?

(Probably related to #104)
(cc: @sruehs)

### PR DESCRIPTION
**Pull request by erikvansebille**
Changing the installation scripts and notes to install via `conda-forge`, as suggested by @willirath and @rabernat. 
Aim is to fix #225 and #104, and also #226 as that was playing havoc on Travis

### CODE REVIEW

**Comment by willirath** [comment]
Here's my suggestion for the “Installing Parcels” section of the documentation:

```rst
Installing Parcels
===================================

Parcels depends on a working Python installation, a netCDF installation, a C
compiler, and various Python packages.  These requirements can be met in
different ways:

1. Use Anaconda and Conda-Forge for all requirements.  This is the
   “batteries-included” solution probably suitable for most users.

2. Use an existing netCDF installation and an existing C compiler and just get
   the Python part of the dependencies.  This may be a suitable solution for
   more experienced users who want to control the details.


Installing Parcels and all dependencies with Anaconda and Conda-Forge
---------------------------------------------------------------------

Installing Anaconda’s Miniconda
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

(If you already have a working Anaconda installation, skip to “Creating an
environment with Parcels”.)

Download Anaconda’s Miniconda for Linux::

    wget https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh \
        -O miniconda.sh

or for MacOSX::

    wget https://repo.continuum.io/miniconda/Miniconda3-latest-MacOSX-x86_64.sh \
        -O miniconda.sh

and install with::

    bash miniconda.sh -b -p $HOME/miniconda


Creating an environment with Parcels
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

With a file `environment.yml`::

    name: py2_parcels
    channels:
      - https://conda.anaconda.org/conda-forge
      - https://repo.continuum.io/pkgs/free
    dependencies:
      - python=2.7
      - basemap
      - cachetools>=1.0.0
      - cgen
      - enum34
      - ffmpeg
      - flake8>=2.1.0
      - gcc
      - jupyter
      - matplotlib=2.0.2
      - netcdf4>=1.1.9
      - numpy>=1.9.1
      - progressbar2
      - py>=1.4.27
      - pymbolic
      - python-dateutil
      - scipy>=0.16.0
      - six>=1.10.0
      - xarray>=0.5.1
      - pip:
          - pytest>=2.7.0
          - nbval
          - "git+https://github.com/OceanParcels/parcels.git@master"

run::

    source $HOME/miniconda/bin/activate root
    conda env create -f environment.yml


Using the Parcels environment
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To start working with Parcels, activate the environment with::

    source $HOME/miniconda/bin/activate py2_parcels


Just installing the python requirements
---------------------------------------

The latest version of Parcels, including tests and examples,
can be obtained directly from github via::

    git clone https://github.com/OceanParcels/parcels.git
    cd parcels
    git checkout master
    pip install -r requirements.txt
    pip install -e .

Note that a functional NetCDF installation and a working C compiler is required.
```

This is rather lengthy and still needs some work to make sure it really reflects the state of the repository (is there / should there still be a `requirements.txt`? etc.).  But maybe it's a good start to discuss how much and how to document the installation process.

File: environment.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+name: py2_parcels
+channels:
+  - https://conda.anaconda.org/conda-forge
+  - https://repo.continuum.io/pkgs/free
+dependencies:
+  - python=2.7
+  - cachetools>=1.0.0
```
Comment (willirath) [inline environment.yml]: I've seen `basemap` mentioned a few times. Should it be included as a requirement here?

Comment (erikvansebille) [reply environment.yml]: We're not enforcing `Basemap`, because on many HPC systems it's not available.  Parcels still runs without `Basemap` installed; users just won't be able to plot continents etc

**Comment by erikvansebille** [comment]
Thanks @willirath for your extended proposal for new Installation guide. I think it looks good in many ways, although I would change a few things:
1. Refer simply to https://conda.io/docs/user-guide/install/ on how to install Miniconda. Advantage is that this might (?) even work on Windows?
2. Simply link to the `environment.yml` file on the master branch of the Parcels repo, rather then spell it out completely
3. Perhaps not even mentioning your option 2 of just installing the python environment? Or, if we mention it, somehow use the same `environment.yml` file? So that we don't have redundancy of both an `environment.yml` file and a `requirements.txt`. Is that even possible?

**Comment by willirath** [comment]
> Refer simply to https://conda.io/docs/user-guide/install/ on how to install Miniconda. Advantage is that this might (?) even work on Windows?

Yes, that's good.  (On Windows, there's no gcc package. So I'm not sure this would lead anywhere.)

> Simply link to the environment.yml file on the master branch of the Parcels repo, rather then spell it out completely

Yes, I also like that.

> Perhaps not even mentioning your option 2 of just installing the python environment? Or, if we mention it, somehow use the same environment.yml file? So that we don't have redundancy of both an environment.yml file and a requirements.txt. Is that even possible?

Yes, option 2 could be dropped entirely. (It's probably safe to assume that anyone who insists on maintaining an own installation of any of the dependencies can figure this out themselves.

I'm not sure about deleting the `requirements.txt` though.

File: docs/index.rst [hunk 2]
```diff
@@ -72,18 +72,31 @@ The best way to get started with Parcels is to have a look at the Jupyter notebo
...
+
+Installing Parcels and all dependencies with Anaconda and Conda-Forge
+-----------------------------------
+
+Install Anaconda's Miniconda following the steps at https://conda.io/docs/user-guide/install/
+
+Download the `environment.yml <https://raw.githubusercontent.com/OceanParcels/parcels/install_via_conda/environment.yml>`_ file and place it in the directory where you want to install Parcels. Then, run::
+
+    source $HOME/miniconda/bin/activate root
+    conda env create -f environment.yml
```
Comment (willirath) [inline docs/index.rst]: I'd use
```bash
conda env create -f environment.yml -n py2_parcels
```
which is redundant but works (`-n` actually overrides the `name:` given in the env file) and makes it obvious why the user is alter asked to activate an env with this name.

File: docs/index.rst [hunk 3]
```diff
@@ -72,18 +72,31 @@ The best way to get started with Parcels is to have a look at the Jupyter notebo
...
+2. Use an existing netCDF installation and an existing C compiler and just get
+   the Python part of the dependencies.  This may be a suitable solution for
+   more experienced users who want to control the details.
+
+Installing Parcels and all dependencies with Anaconda and Conda-Forge
+-----------------------------------
+
+Install Anaconda's Miniconda following the steps at https://conda.io/docs/user-guide/install/
+
+Download the `environment.yml <https://raw.githubusercontent.com/OceanParcels/parcels/install_via_conda/environment.yml>`_ file and place it in the directory where you want to install Parcels. Then, run::
```
Comment (willirath) [inline docs/index.rst]: I think the "place it in the directory where you want to install Parcels" is misleading as the `environment.yml` only has to be present while the environment is created. Would it be still understandable, if the whole thing would be truncated to
> Download the ... file and run::

Another thing to keep in mind: The link should point to
<https://raw.githubusercontent.com/OceanParcels/parcels/master/environment.yml>
in the end.

File: docs/index.rst (hunk 2 above)
Comment (willirath) [inline docs/index.rst]: Another thing:  The installation is only completed after using pip to install from the `master` branch of Parcels.  So the whole code block is:
```bash
source $HOME/miniconda/bin/activate root
conda env create -f environment.yml -n py2_parcels
pip install git+https://github.com/OceanParcels/parcels.git@master
```

File: environment.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+name: py2_parcels
+channels:
+  - https://conda.anaconda.org/conda-forge
+  - https://repo.continuum.io/pkgs/free
```
Comment (willirath) [inline environment.yml]: The channels field can be reduced to 
```yaml
channels:
- conda-forge
```

The second line is the default fall-back channel for anything not found in explicitly listed channels anyway. And the explicit URLs might change in the future while the short-form `conda-forge` will just refer to whatever is the current conda-forge-repo URL. (I tend to put full URLs there, because I have some setups behind firewalls where I have to point this to a mirror.)

File: docs/index.rst (hunk 3 above)
Comment (erikvansebille) [reply docs/index.rst]: Good point. I've now also added a check of broken files in the documentation using the `sphinx linkcheck` functionality. So even if we forget, travis will complain

File: docs/index.rst [hunk 5]
```diff
@@ -78,7 +78,7 @@ compiler, and various Python packages.
 The simplest way to install Parcels is to use Anaconda and Conda-Forge for all 
 requirements. This is the “batteries-included” solution probably suitable for most users.
 
-1. Install Anaconda's Miniconda following the steps at https://conda.io/docs/user-guide/install/
+1. Install Anaconda's Miniconda following the steps at https://conda.io/docs/user-guide/install/. Be sure to download the Python 2.7 version.
```
Comment (willirath) [inline docs/index.rst]: I'd rephrase this to:

> All the code below assumes that you download the Python-2 version.

Here's why: The Python version for the Parcels environment is specified in the `environment.yml`. Conda is perfectly fine with different Python versions in different envs. I usually have a Miniconda3 root env with different Python-2 and Python-3 environments for different use cases.

File: docs/index.rst [hunk 6]
```diff
@@ -69,26 +69,45 @@ The best way to get started with Parcels is to have a look at the Jupyter notebo
...
-    git clone https://github.com/OceanParcels/parcels.git
-    cd parcels; pip install -r requirements.txt
-    python scripts/pull_data.py
-    export PYTHONPATH="$PYTHONPATH:$PWD"
+5. The next time you log in to your machine and want to work with Parcels, activate the environment with::
 
-In order for Parcels to work from any directory, add the following line to 
-your ~/.bash_profile::
-    export PYTHONPATH="$PYTHONPATH:$PWD"
+    source $HOME/miniconda2/bin/activate py2_parcels
```
Comment (erikvansebille) [inline docs/index.rst]: @willirath is there a reason you suggested to `source $HOME/miniconda2/bin/activate py2_parcels` whereas in most conda documentation and tutorials people use `source activate py2_parcels`?

Comment (willirath) [reply docs/index.rst]: The short version needs the Conda's `activate` executable in the Path.  During the Miniconda installation, the users have to decide if they want to add the installation to the Path. It is debatable whether this is a good idea or not.  (I prefer not to.)

The long version which explicitly includes the full path to `activate`, is working in any case.

File: setup.cfg [hunk 7]
```diff
@@ -1,5 +1,5 @@
 [flake8]
-ignore = E501,F403,E226,W503
+ignore = E501,F403,E226,E722,W503
```
Comment (delandmeterp) [inline setup.cfg]: Should we also ignore error E731?
It makes the following error in another branch
"test_fieldset.py:186:5: E731 do not assign a lambda expression, use a def"

File: docs/index.rst (hunk 6 above)
Comment (delandmeterp) [reply docs/index.rst]: Ok with the source comment, but you changed here the meaning of the sentence.
It was already stated that user has to source the activate file (line 87).
The point here is to say that the source can be written in the .bash_profile

Comment (erikvansebille) [reply docs/index.rst]: Well, but do we really want people to source the Parcels environment in their `.bash_rc`? That goes against the idea of conda environments, where you only source them if you need them. So I prefer not to suggest adding to `.bash_rc`

File: setup.cfg (hunk 7 above)
Comment (erikvansebille) [reply setup.cfg]: I didn't realise that using lambda functions is against PEP8, but actually that does make sense. Why not use a `def` indeed, for readability, in PR #228?

File: docs/index.rst (hunk 6 above)
Comment (delandmeterp) [reply docs/index.rst]: Ok, you're right

File: setup.cfg (hunk 7 above)
Comment (delandmeterp) [reply setup.cfg]: I'll do the change in the other branch.
